### DISCUSSION TITLE
PI4J LedStrip

### DISCUSSION THREAD

**Discussion by livansilvagarcia** [Pi4J/pi4j#367] [Q&A]
Hello everyone, I am testing Pi4j on a Pi Zero 2 W board, but the led strip is not working. I am using version 2.6 of Pi4j. I have imported all the classes from the [components ](https://github.com/Pi4J/pi4j-example-components/blob/main/src/main/java/com/pi4j/catalog/components/LedStrip.java)package in my project. 

This is the code

public class AC_1 {

    public static void main(String[] args) throws InterruptedException {
        
        System.out.println("LED strip app started ...");

        Context pi4j = Pi4J.newAutoContext();

        // Initialize the strip
        Duration ms50 = Duration.ofMillis(50);
        Duration ms500 = Duration.ofMillis(500);
        Duration sec1 = Duration.ofSeconds(1);

        int leds = 10;
        final LedStrip ledStrip = new LedStrip(pi4j, leds);

        Thread.sleep(1000);

        Index.getInstance().setVisible(true);

        while (true) {

            System.out.println("LED strip shines purple");
            Index.getInstance().upDateView("LED strip shines purple");
            ledStrip.setStripColor(LedStrip.LedColor.PURPLE);
            ledStrip.render(sec1);

            System.out.println("turn strip off");
            Index.getInstance().upDateView("turn strip off");
            ledStrip.allOff();
            ledStrip.render(sec1);

            System.out.println("toggle between GREEN and RED");
            Index.getInstance().upDateView("toggle between GREEN and RED");
            ledStrip.alternate(LedStrip.LedColor.GREEN, LedStrip.LedColor.RED, ms500, 3);

            System.out.println("setting the LEDs to blue and the first one to purple");
            Index.getInstance().upDateView("setting the LEDs to blue and the first one to purple");
            ledStrip.setStripColor(LedStrip.LedColor.BLUE);
            ledStrip.setPixelColor(0, LedStrip.LedColor.PURPLE);
            ledStrip.render(ms500);

            System.out.println("Start a kind of animation");
            Index.getInstance().upDateView("Start a kind of animation");
            for (int i = 0; i < leds - 1; i++) {
                ledStrip.setPixelColor(i, LedStrip.LedColor.BLUE);
                ledStrip.setPixelColor(i + 1, LedStrip.LedColor.PURPLE);
                ledStrip.render(ms50);
            }

            for (int i = leds - 1; i > 0; i--) {
                ledStrip.setPixelColor(i, LedStrip.LedColor.BLUE);
                ledStrip.setPixelColor(i - 1, LedStrip.LedColor.PURPLE);
                ledStrip.render(ms50);
            }

            Thread.sleep(100);

            System.out.println("setting the brightness to full and show the first LED as white");
            Index.getInstance().upDateView("setting the brightness to full and show the first LED as white");
            ledStrip.allOff();
            ledStrip.setMaxBrightness(1);
            ledStrip.setPixelColor(0, LedStrip.LedColor.WHITE);
            ledStrip.render(Duration.ofSeconds(2));

            //finishing and closing
            ledStrip.reset();

            System.out.println("LED strip demo finished.");
            Index.getInstance().upDateView("LED strip demo finished.");
        }

    }
}

**Comment by FDelporte** [comment]
Hi, LED strips are a very hard component to control from a Linux-based system because of the exact timing that is required to control these. Did you take this remarks in the original code into account?
https://github.com/Pi4J/pi4j-example-components/blob/main/src/main/java/com/pi4j/catalog/components/LedStrip.java#L13

A better approach is to use an external device that is easier to communicate with, e.g. the Pixelblaze via serial communication, as described here: https://www.pi4j.com/examples/jbang/pixelblaze_output_expander/
